torch/lib/c10d/FileStore.cpp

```diff
@@ -0,0 +1,284 @@
+    if (rv < 0) {
+      throw std::system_error(errno, std::generic_category(), "flock");
+    }
+  }
+};
+
+class File {
+ public:
+  explicit File(const std::string& path, int flags) {
+    fd_ = syscall(std::bind(::open, path.c_str(), flags, 0644));
+    if (fd_ < 0) {
+      throw std::system_error(
+          errno,
+          std::generic_category(),
+          "Unable to open " + path);
+    }
+  }
+
+  ~File() {
+    ::close(fd_);
+  }
+
+  Lock lockShared() {
+    return Lock(fd_, LOCK_SH);
+  }
+
+  Lock lockExclusive() {
+    return Lock(fd_, LOCK_EX);
+  }
+
+  off_t seek(off_t offset, int whence) {
+    auto rv = syscall(std::bind(lseek, fd_, offset, whence));
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t tell() {
+    auto rv = syscall(std::bind(lseek, fd_, 0, SEEK_CUR));
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t size() {
+    auto pos = tell();
+    auto size = seek(0, SEEK_END);
+    seek(pos, SEEK_SET);
+    return size;
+  }
+
+  void write(const void* buf, size_t count) {
+    while (count > 0) {
+      auto rv = syscall(std::bind(::write, fd_, buf, count));
+      if (rv < 0) {
+        throw std::system_error(errno, std::system_category());
+      }
+      buf = (uint8_t*) buf + count;
+      count -= rv;
+    }
+  }
+
+  void read(void* buf, size_t count) {
+    while (count > 0) {
+      auto rv = syscall(std::bind(::read, fd_, buf, count));
+      if (rv < 0) {
+        throw std::system_error(errno, std::system_category());
+      }
+      buf = (uint8_t*) buf + count;
+      count -= rv;
+    }
+  }
+
+  void write(const std::string& str) {
+    uint32_t len = str.size();
+    write(&len, sizeof(len));
+    write(str.c_str(), len);
+  }
+
+  void write(const std::vector<uint8_t>& data) {
+    uint32_t len = data.size();
+    write(&len, sizeof(len));
+    write(data.data(), len);
+  }
+
+  void read(std::string& str) {
+    uint32_t len;
+    read(&len, sizeof(len));
+    std::vector<uint8_t> buf(len);
+    read(buf.data(), len);
+    str.assign(buf.begin(), buf.end());
+  }
+
+  void read(std::vector<uint8_t>& data) {
+    uint32_t len;
+    read(&len, sizeof(len));
+    data.resize(len);
+    read(data.data(), len);
+  }
+
+ protected:
+  int fd_;
+};
+
+off_t refresh(
+    File& file,
+    off_t pos,
+    std::map<std::string, std::vector<uint8_t>>& cache) {
+  auto size = file.size();
+  if (size != pos) {
+    std::string tmpKey;
+    std::vector<uint8_t> tmpValue;
+    file.seek(pos, SEEK_SET);
+    while (size > pos) {
+      file.read(tmpKey);
+      file.read(tmpValue);
+      cache[tmpKey] = std::move(tmpValue);
+      pos = file.tell();
+    }
+  }
+  return pos;
+}
+
+} // namespace
+
+FileStore::FileStore(const std::string& path)
+    : Store(),
+      path_(path),
+      pos_(0) {
+}
+
+FileStore::~FileStore() {
+}
+
+void FileStore::set(const std::string& key, const std::vector<uint8_t>& value) {
+  File file(path_, O_RDWR | O_CREAT);
```

Comment:
Actually is there a reason why we want to reopen the file at every operation? It might be fairly expensive if you're using network file systems I think. We could just cache this RDWR object in `FileStore`